# **Project Name**    -  **Play Store EDA**

##### **Project Type**    - EDA
##### **Contribution**    - Individual
##### **Rakshit Pandey**

# **Project Summary -**

Write the summary here within 500-600 words.

### Introduction:
The Play Store apps data provides valuable insights for app-making businesses aiming for success in the Android market. Understanding user preferences and factors driving app engagement is crucial for developers to improve app quality and user satisfaction. In this exploratory data analysis (EDA) project, we aim to delve into the Play Store apps dataset and customer reviews to uncover key insights that can guide app development strategies and enhance user experience.

### Dataset Description:
The dataset comprises customer reviews of Android apps, containing fields such as review ID, username, review text, rating, upvotes, version, and timestamps. Each review offers valuable feedback from users, providing insights into user sentiment, satisfaction, and areas for improvement.

### Objectives:

1. **Explore User Reviews**: Analyze user reviews to understand sentiment and identify common themes.
2. **Investigate App Ratings**: Examine the distribution of app ratings and identify factors influencing positive and negative ratings.
3. **Discover Trends Over Time**: Explore trends in review counts, app versions, and user engagement.
4. **Compare Free vs. Paid Apps**: Compare the performance and user sentiment of free and paid apps.
5. **Visualize Insights**: Present key insights through various visualizations to aid in interpretation and decision-making.

### Methodology:

1. **Data Loading and Cleaning**: Load the dataset and perform data cleaning to handle missing values and inconsistencies.
2. **Understanding Variables**: Explore dataset columns, summary statistics, and unique values to understand the variables.
3. **Data Wrangling**: Prepare the data for analysis by performing necessary manipulations and transformations.
4. **Data Visualization and Storytelling**:
   - Create visualizations to understand relationships between variables and uncover insights.
   - Utilize storytelling to interpret findings and draw actionable insights.
5. **Key Findings and Recommendations**: Summarize key findings and provide recommendations for app developers and stakeholders.


### Key Findings:

    Key findings will come here 

---

# **GitHub Link -**

Provide your GitHub Link here.

https://github.com/pandey-rakshit/almabetter/tree/main/Module2/capstone/Play_store_EDA

# **Problem Statement**


**Write Problem Statement Here.**

### Business Context:
The Android app market is highly competitive, with millions of apps available on the Google Play Store. For app developers and businesses, understanding user preferences and factors influencing app engagement is crucial for success. User reviews play a significant role in shaping the reputation and popularity of apps, while app ratings provide valuable feedback on user satisfaction. However, navigating through vast amounts of data to extract meaningful insights can be challenging.

### Problem:
The challenge is to analyze the Play Store apps dataset and customer reviews to uncover key factors responsible for app engagement and success. The primary goal is to provide actionable insights for developers to improve their apps and enhance user satisfaction. Specific objectives include:

1. **Understanding User Sentiment**: Analyze user reviews to understand sentiment and identify common themes, issues, and preferences.
2. **Investigating App Ratings**: Examine the distribution of app ratings and identify factors influencing positive and negative ratings.
3. **Discovering Trends Over Time**: Explore trends in review counts, app versions, and user engagement to identify patterns and fluctuations.
4. **Comparing Free vs. Paid Apps**: Compare the performance and user sentiment of free and paid apps to understand their impact on user satisfaction.
5. **Visualizing Insights**: Present key insights through various visualizations to aid in interpretation and decision-making.


#### **Define Your Business Objective?**

Answer Here.

The ultimate goal is to provide actionable insights for app developers and stakeholders to improve app quality, user experience, and overall success in the Android app market. By understanding user sentiment, app ratings, and trends over time, developers can make informed decisions to enhance their apps, attract more users, and increase user satisfaction.

# **General Guidelines** : -  

1.   Well-structured, formatted, and commented code is required. 
2.   Exception Handling, Production Grade Code & Deployment Ready Code will be a plus. Those students will be awarded some additional credits. 
     
     The additional credits will have advantages over other students during Star Student selection.
       
             [ Note: - Deployment Ready Code is defined as, the whole .ipynb notebook should be executable in one go
                       without a single error logged. ]

3.   Each and every logic should have proper comments.
4. You may add as many number of charts you want. Make Sure for each and every chart the following format should be answered.
        

```
# Chart visualization code
```
            

*   Why did you pick the specific chart?
*   What is/are the insight(s) found from the chart?
* Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

5. You have to create at least 20 logical & meaningful charts having important insights.


[ Hints : - Do the Vizualization in  a structured way while following "UBM" Rule. 

U - Univariate Analysis,

B - Bivariate Analysis (Numerical - Categorical, Numerical - Numerical, Categorical - Categorical)

M - Multivariate Analysis
 ]





# ***Let's Begin !***

## Dependencies

### Import Libraries

In [1]:
# Import Libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Dataset Loading

In [2]:
# Load Dataset

def load_dataset(file_path):
    """
    Load the dataset from the given file path.
    
    Args:
    - file_path (str): Path to the CSV file
    
    Returns:
    - df (DataFrame): Loaded DataFrame
    """
    try:
        df = pd.read_csv(file_path)
        return df
    except FileNotFoundError:
        print("File not found. Please provide the correct file path.")
        return None

### Dataset Wrangling

#### Handling Missing Values

In [3]:
# handling missing values

def handle_missing_values(df):
    """
    Handle missing values in the DataFrame.
    
    Args:
    - df (DataFrame): Input DataFrame
    
    Returns:
    - df (DataFrame): DataFrame with missing values handled
    """
    # Check missing values
    print(f"Missing values before handling: \n{df.isnull().sum()}\n")
    
    # Drop rows with missing values
    df.dropna(inplace=True)
    
    # Check missing values after handling
    print(f"\nMissing values after handling: \n{df.isnull().sum()}\n")
    
    return df


#### Handling Duplicates Values

In [4]:
# Handling Duplicates Values

def handle_duplicates(df):
    """
    Handle duplicate rows in the DataFrame.
    
    Args:
    - df (DataFrame): Input DataFrame
    
    Returns:
    - df (DataFrame): DataFrame with duplicates removed
    """
    # Check duplicate rows
    print(f"Duplicate rows before handling: \n{df.duplicated().sum()}\n")
    
    # Remove duplicate rows
    df.drop_duplicates(inplace=True)
    
    # Check duplicate rows after handling
    print(f"Duplicate rows after handling: \n{df.duplicated().sum()}\n")
    
    return df

#### Datatype casting

In [5]:
# convert data types | datatype casting

def convert_data_types(df, column_dtype_map):
    """
    Convert columns to specified data types in the DataFrame.
    
    Args:
    - df (DataFrame): Input DataFrame
    - column_dtype_map (dict): Dictionary mapping column names to target data types
    
    Returns:
    - df (DataFrame): DataFrame with converted data types
    """
    # Iterate through each column and its target data type
    for column, dtype in column_dtype_map.items():
        # Convert column to specified data type
        df[column] = df[column].astype(dtype)
    
    return df

### Data Visualization 

In [6]:
# chart visualzation function
def visualize_chart(plot_function, title, xlabel, ylabel, x, y):
    """
    Create a custom visualization chart.
    
    Parameters:
    - plot_function (function): Function for plotting, e.g., sns.scatterplot, sns.lineplot, sns.barplot.
    - title (str): Title of the chart.
    - xlabel (str): Label for the x-axis.
    - ylabel (str): Label for the y-axis.
    - x (array-like): Data for the x-axis.
    - y (array-like): Data for the y-axis.
    
    Returns:
    - fig (matplotlib.figure.Figure): The created figure object.
    """
    plt.figure(figsize=(10, 6))
    
    plot_function(x=x, y=y)
    
    plt.title(title)
    plt.xlabel(xlabel)
    plt.ylabel(ylabel)
    plt.xticks(rotation=45)
    
    fig = plt.gcf()
    
    return fig


In [7]:
# Define a dictionary to map chart types to plotting functions
plot_functions = {
    'scatter': sns.scatterplot,
    'line': sns.lineplot,
    'bar': sns.barplot
}

# Play Store Data Analysis

## ***1. Know Your Data*** 

### Read Data File

In [8]:
# read_csv file "Play Store Data.csv"
play_store_file = 'https://raw.githubusercontent.com/pandey-rakshit/almabetter/main/Module2/capstone/Play_store_EDA/data/Play%20Store%20Data.csv'

# loading dataset - play_store_df : DataFrame
play_store_df = load_dataset(play_store_file)

### Dataset First View

In [9]:
# Dataset First Look
play_store_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [10]:
# Renaming df columns: Replacing space by underscore "_"
play_store_df.columns = play_store_df.columns.str.replace(' ', '_')
play_store_df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Current_Ver',
       'Android_Ver'],
      dtype='object')

### Dataset Rows & Columns count

In [11]:
# Dataset Rows & Columns count
play_store_df.shape

(10841, 13)

### Dataset Information

In [12]:
# Dataset Info
play_store_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content_Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last_Updated    10841 non-null  object 
 11  Current_Ver     10833 non-null  object 
 12  Android_Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


#### Duplicate Values

In [13]:
# Dataset Duplicate Value Count

play_store_df.duplicated().sum()

483

#### Missing Values/Null Values

In [14]:
# Missing Values/Null Values Count

play_store_missing_value_counts = play_store_df.isna().sum()
play_store_missing_value_counts

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content_Rating       1
Genres               0
Last_Updated         0
Current_Ver          8
Android_Ver          3
dtype: int64

In [ ]:
# Visualizing the missing values

title = 'Missing Values Chart : Play Store Data'
xlabel = 'Dataset Columns'
ylabel = 'Missing Values Count'
x = play_store_missing_value_counts.index
y = play_store_missing_value_counts.values
fig = visualize_chart(plot_functions['bar'], title, xlabel, ylabel, 
x, y)

### What did you know about your dataset?

Answer Here

#### Play Store Data

1. **Size and Structure**:
   - The DataFrame has 10,841 rows and 13 columns.
   - Each row represents an entry (or record) about a mobile app.
   - The columns are:
     - App: Name of the application.
     - Category: Category of the application.
     - Rating: Average user rating of the application.
     - Reviews: Number of user reviews for the application.
     - Size: Size of the application.
     - Installs: Number of downloads/installs of the application.
     - Type: Whether the application is free or paid.
     - Price: Price of the application.
     - Content Rating: Content rating of the application.
     - Genres: Genre(s) of the application.
     - Last Updated: Date when the application was last updated.
     - Current Ver: Current version of the application.
     - Android Ver: Minimum required Android version.

2. **Data Types and Missing Values**:
   - The 'Rating' column is of float type.
   - The 'Reviews', 'Size', 'Installs', 'Price', 'Last Updated', 'Current Ver', and 'Android Ver' columns are of object type.
   - There are missing values in the DataFrame:
     - 'Rating' column has 1474 missing values.
     - 'Type', 'Content Rating', 'Current Ver', and 'Android Ver' columns have 1, 1, 8, and 3 missing values respectively.

3. **Duplicates**:
   - There are 483 duplicate entries in the DataFrame.


## ***2. Understanding Your Variables***

In [15]:
# Dataset Columns
play_store_df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Current_Ver',
       'Android_Ver'],
      dtype='object')

In [16]:
# Dataset Describe
play_store_df.describe()

Rating
count  9367.000000
mean      4.193338
std       0.537431
min       1.000000
25%       4.000000
50%       4.300000
75%       4.500000
max      19.000000

### Variables Description 

Answer Here

#### Play Store Data

1. **Outliers**:
   - An outlier with a rating of 19 was identified, indicating potential data entry errors or data quality issues.
   - It is important to investigate and potentially correct or remove this outlier to avoid skewing the analysis.

2. **Missing Values**:
   - The 'Rating' column has 1474 missing values.
   - Missing values should be handled appropriately, either by imputation or by removing rows with missing values.
   - Addressing missing values is crucial for accurate analysis and interpretation.

3. **Duplicated Data**:
   - There are 483 duplicated entries in the dataset.
   - Duplicates need to be removed to prevent overcounting and ensure accuracy in the analysis.

4. **Summary**:
   - The 'Rating' column has an average rating of approximately 4.19 out of 5, with a moderate level of variability indicated by a standard deviation of around 0.54.
   - The presence of an outlier with a rating of 19 and missing values in the 'Rating' column, along with duplicated entries, need to be addressed for accurate analysis.
   - Addressing outliers, missing values, and duplicates is crucial for a reliable interpretation of the data.

### Check Unique Values for each variable.

In [17]:
# Check Unique Values for each variable. -- Play Store Data
play_store_df.nunique()

App               9660
Category            34
Rating              40
Reviews           6002
Size               462
Installs            22
Type                 3
Price               93
Content_Rating       6
Genres             120
Last_Updated      1378
Current_Ver       2832
Android_Ver         33
dtype: int64

## 3. ***Data Wrangling***

### Data Wrangling Code

In [ ]:
# Write your code to make your dataset analysis ready.

### What all manipulations have you done and insights you found?

Answer Here.

## ***4. Data Vizualization, Storytelling & Experimenting with charts : Understand the relationships between variables***

#### Chart - 1

In [ ]:
# Chart - 1 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 2

In [ ]:
# Chart - 2 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 3

In [ ]:
# Chart - 3 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 4

In [ ]:
# Chart - 4 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 5

In [ ]:
# Chart - 5 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 6

In [ ]:
# Chart - 6 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 7

In [ ]:
# Chart - 7 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 8

In [ ]:
# Chart - 8 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 9

In [ ]:
# Chart - 9 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 10

In [ ]:
# Chart - 10 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 11

In [ ]:
# Chart - 11 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 12

In [ ]:
# Chart - 12 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 13

In [ ]:
# Chart - 13 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact? 
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 14 - Correlation Heatmap

In [ ]:
# Correlation Heatmap visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

#### Chart - 15 - Pair Plot 

In [ ]:
# Pair Plot visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

# **Solution to Business Objective**

#### What do you suggest the client to achieve Business Objective ? 
Explain Briefly.

Answer Here.

# **Conclusion**

Write the conclusion here.

### ***Hurrah! You have successfully completed your EDA Capstone Project !!!***